src/graph.rs

```diff
@@ -186,11 +186,9 @@ impl<'hw: 'op, 'op> Graph<'hw, 'op> {
     /// # Returns
     ///
     /// * Calculated/cached value associated to `target`.
-    pub(crate) fn calculate(&mut self, target: usize) -> Result<Array<'hw>> {
+    pub(crate) fn calculate(&mut self, target: usize) -> &Array<'hw> {
         // Avoiding an edge case: inner step_ids should be correct, but `target` is not constrained.
-        if target >= self.steps.len() {
-            return Err(Error::InvalidNode(format!("Invalid step ID: {}", target)));
-        }
+        assert!(target < self.steps.len(), "Invalid step ID: {}", target);
```

Comment:
Could you consider removing this assertion and using `get()` and `expect()` instead of `get_unchecked()` in line 252?

It is better not to use `unsafe` if it is equivalent.

Reply:
I understand. The push-down automaton depends on `target` and the value must be checked before it.